lib/icingadb/redisconnection.cpp

```diff
@@ -1,85 +1,115 @@
+RedisConnection::RedisConnection(asio::io_context& io, const ObjectImpl<IcingaDB>::ConstPtr& icingadb, const Ptr& parent, bool trackOwnPendingQueries)
+	: m_IcingaDB{icingadb}, m_Strand(io), m_Connecting(false), m_Connected(false), m_Stopped(false),
+	m_QueuedWrites(io), m_QueuedReads(io), m_TrackOwnPendingQueries{trackOwnPendingQueries}, m_LogStatsTimer(io),
+	m_Parent(parent)
 {
-	if (useTls && m_Path.IsEmpty()) {
+	if (icingadb->GetEnableTls() && icingadb->GetPath().IsEmpty()) {
 		UpdateTLSContext();
 	}
 }
 
 void RedisConnection::UpdateTLSContext()
 {
-	m_TLSContext = SetupSslContext(m_CertPath, m_KeyPath, m_CaPath,
-		m_CrlPath, m_CipherList, m_TlsProtocolmin, m_DebugInfo);
+	m_TLSContext = SetupSslContext(
+		m_IcingaDB->GetCertPath(),
+		m_IcingaDB->GetKeyPath(),
+		m_IcingaDB->GetCaPath(),
+		m_IcingaDB->GetCrlPath(),
+		m_IcingaDB->GetCipherList(),
+		m_IcingaDB->GetTlsProtocolmin(),
+		m_IcingaDB->GetDebugInfo()
+	);
 }
 
 void RedisConnection::Start()
 {
-	if (!m_Started.exchange(true)) {
-		Ptr keepAlive (this);
+	ASSERT(!m_Connected && !m_Connecting);
 
-		IoEngine::SpawnCoroutine(m_Strand, [this, keepAlive](asio::yield_context yc) { ReadLoop(yc); });
-		IoEngine::SpawnCoroutine(m_Strand, [this, keepAlive](asio::yield_context yc) { WriteLoop(yc); });
+	Ptr keepAlive (this);
+	m_Stopped.store(false);
 
-		if (!m_Parent) {
-			IoEngine::SpawnCoroutine(m_Strand, [this, keepAlive](asio::yield_context yc) { LogStats(yc); });
-		}
-	}
+	IoEngine::SpawnCoroutine(m_Strand, [this, keepAlive](asio::yield_context yc) { ReadLoop(yc); });
+	IoEngine::SpawnCoroutine(m_Strand, [this, keepAlive](asio::yield_context yc) { WriteLoop(yc); });
 
-	if (!m_Connecting.exchange(true)) {
-		Ptr keepAlive (this);
-
-		IoEngine::SpawnCoroutine(m_Strand, [this, keepAlive](asio::yield_context yc) { Connect(yc); });
+	if (!m_Parent) {
+		IoEngine::SpawnCoroutine(m_Strand, [this, keepAlive](asio::yield_context yc) { LogStats(yc); });
 	}
+
+	m_Connecting.store(true);
+	IoEngine::SpawnCoroutine(m_Strand, [this, keepAlive](asio::yield_context yc) { Connect(yc); });
 }
 
-bool RedisConnection::IsConnected() {
-	return m_Connected.load();
+/**
+ * Disconnect the Redis connection gracefully.
+ *
+ * This function initiates a graceful disconnection of the Redis connection. It sets the stopped flag to true,
+ * and spawns a coroutine to handle the disconnection process. The coroutine waits for up to 5 seconds for any
+ * ongoing operations to finish before proceeding with the disconnection.
+ */
+void RedisConnection::Disconnect()
+{
+	if (!m_Stopped.exchange(true)) {
+		IoEngine::SpawnCoroutine(m_Strand, [this, keepAlive = Ptr(this)](asio::yield_context yc) {
+			// Wait up to 5 seconds for ongoing operations to finish.
+			asio::deadline_timer waiter(m_Strand.context(), boost::posix_time::seconds(5));
+			waiter.async_wait(yc);
+
+			m_QueuedWrites.Set(); // Wake up write loop
+			m_QueuedReads.Set(); // Wake up read loop
```

Comment:
Wait up to 5 seconds? Doesn't this just always wait 5 seconds?

And why are the read/write loops only woken after that?

Reply:
> Wait up to 5 seconds? Doesn't this just always wait 5 seconds?

Yes, it always waits 5 seconds.

> And why are the read/write loops only woken after that?

Doesn't make any difference but I've moved these above the timer now.